xwiki-platform-core/xwiki-platform-refactoring/xwiki-platform-refactoring-api/src/main/java/org/xwiki/refactoring/internal/job/AbstractCopyOrMoveJob.java

```diff
@@ -121,6 +157,71 @@ private boolean isDescendantOrSelf(EntityReference alice, EntityReference bob)
+            checkSourceDestination(source, destination);
+            DocumentReference destinationDocumentReference;
+            if (processOnlySameSourceDestinationTypes()) {
+                putInConcernedEntitiesOnlySameSource(source, destination);
+            } else {
+                if (this.request.isDeep() && isSpaceHomeReference(source)) {
+                    visitSpace(source.getLastSpaceReference(), destination, this::putInConcernedEntities);
+                } else if (destination.getType() == EntityType.SPACE) {
+                    destinationDocumentReference =
+                        new DocumentReference(source.getName(), new SpaceReference(destination));
+                    this.putInConcernedEntities(source, destinationDocumentReference);
+                } else if (destination.getType() == EntityType.DOCUMENT
+                    && isSpaceHomeReference(new DocumentReference(destination))) {
+                    destinationDocumentReference =
+                        new DocumentReference(source.getName(), new SpaceReference(destination.getParent()));
+                    this.putInConcernedEntities(source, destinationDocumentReference);
+                } else if (destination.getType() == EntityType.WIKI) {
+                    visitSpace(source.getLastSpaceReference(), destination, this::putInConcernedEntities);
+                }
+            }
+        } catch (InternalCopyOrMoveJobException e) {
+            this.logger.debug(e.getMessage());
+        }
+    }
+
+    private void putInConcernedEntitiesOnlySameSource(DocumentReference source, EntityReference destination)
+    {
+        DocumentReference destinationDocumentReference = new DocumentReference(destination);
+        if (this.request.isDeep() && isSpaceHomeReference(source)) {
+            if (isSpaceHomeReference(destinationDocumentReference)) {
+                visitSpace(source.getLastSpaceReference(), destinationDocumentReference.getLastSpaceReference(),
+                    this::putInConcernedEntities);
+            }
+        } else {
+            this.putInConcernedEntities(source, destinationDocumentReference);
+        }
+    }
+
+    private void putInConcernedEntities(DocumentReference sourceDocument, DocumentReference destination)
+    {
+        try {
+            if (!this.modelBridge.exists(sourceDocument)) {
+                this.logger.warn("Skipping [{}] because it doesn't exist.", sourceDocument);
+            } else if (this.checkAllRights(sourceDocument, destination)) {
+                this.concernedEntities.put(sourceDocument, new EntitySelection(sourceDocument, destination));
+            }
+        } catch (Exception e) {
+            logger.error("Failed to perform the refactoring from document with reference [{}] to [{}]",
+                sourceDocument, destination, e);
+        }
+    }
+
+    // FIXME: this should be factorized with the code from getTargetReference
```

Comment:
Note: I'll drop that FIXME for after the PR is merged. I'd like to improve the code in AbstractCopyOrMoveJob but it could be done later probably (e.g. between RC and final release).